### DISCUSSION TITLE
Unable to run debug code inside container

### DISCUSSION THREAD

**Discussion by robs-nice99** [microsoft/vscode-docker#3750] [Q&A]
Hello,

I have a python project, running inside a docker, that I'm to debug using VScode, but couldn't (after reading the documentation: [setup](https://code.visualstudio.com/docs/containers/quickstart-python) and [debug](https://code.visualstudio.com/docs/containers/debug-python))
I create a Hello World project to share my problems here.

This is the script 'hw.py':

     print("Hello word!")

After creating this file and open the folder in VSCode, I ran "Docker: Add Docker Files to workspace", which created the expected files (`tasks.json`, `lauch.json`, `Dockerfile` and `requirements.txt`).

However, when I go to "Run and Debug" and try to debug the new "Docker: Python - General" I get the following message:

![image](https://user-images.githubusercontent.com/47054603/205398057-d2afcfd1-3f2f-48f5-b33d-860a622d7b52.png)

I try the several options, without success.

If I try to launch the debugger again, the message no longer appears, but I see this in the output:

*  Executing task: docker-run: debug 

```
> docker container run --detach --tty --name 'pythonindocker-dev' --publish-all --mount 'type=bind,source=/home/fernando.santos/.vscode/extensions/ms-python.python-2022.18.2/pythonFiles/lib/python/debugpy,destination=/debugpy,readonly' --label 'com.microsoft.created-by=visual-studio-code' --entrypoint 'python3' pythonindocker:latest <

91e5e97e333e16cfee623cf9769568febe19df72f3e80d5f0e2b2db9446a0a3etime="2022-12-02T22:08:25Z" level=error msg="error waiting for container: context canceled"
```

And nothing happens (the code does not run).

I also noticed that the created Dockerfile has no exposed port. How is the debugger supposed to communicate with VSCode?

Dockerfile:

```
# For more information, please refer to https://aka.ms/vscode-docker-python
FROM python:3.8-slim

# Keeps Python from generating .pyc files in the container
ENV PYTHONDONTWRITEBYTECODE=1

# Turns off buffering for easier container logging
ENV PYTHONUNBUFFERED=1

# Install pip requirements
COPY requirements.txt .
RUN python -m pip install -r requirements.txt

WORKDIR /app
COPY . /app

# Creates a non-root user with an explicit UID and adds permission to access the /app folder
# For more info, please refer to https://aka.ms/vscode-docker-python-configure-containers
RUN adduser -u 5678 --disabled-password --gecos "" appuser && chown -R appuser /app
USER appuser

# During debugging, this entry point will be overridden. For more information, please refer to https://aka.ms/vscode-docker-python-debug
CMD ["python", "hw.py"]
```

What am I missing?

**Comment by brettcannon** [comment]
Are you using our [dev container](https://code.visualstudio.com/docs/devcontainers/containers) support to develop _within_ the container, or are you explicitly trying to do a remote attach to the debugger running within the container? If you do the former then there's no need to open any ports as VS Code will be running _inside_ the container itself (and will launch the container so VS Code can connect to it).

**Reply by robs-nice99** [reply]
How can I check that? As I wrote above, I followed the "Python in a container" tutorial. It has detailed steps, but doesn't really explain what's happening.

**Reply by brettcannon** [reply]
If you're following that tutorial then you're using https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-docker and not https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers . I will transfer this discussion to the vscode-docker repo as that's where the error is coming from.

**Reply by robs-nice99** [reply]
Thanks @brettcannon

**Comment by bwateratmsft** [comment]
I'm not sure where this `setAtlas` error is coming from, I've never seen it before. It's definitely not the Docker extension, and doesn't look to be from the Python extension or `debugpy` either. I found one relevant [VSCode issue](https://github.com/microsoft/vscode/issues/166909).

> And nothing happens (the code does not run).

I see an error after the `docker run` command, `time="2022-12-02T22:08:25Z" level=error msg="error waiting for container: context canceled"`. I haven't seen that error without any additional accompanying error information. Does the container stay running or immediately quit and/or get removed?

Does the "Python Debug Console" open? Looks like this: 
![image](https://user-images.githubusercontent.com/36966225/206212319-a8193774-a3ae-4ef7-87ca-3b66e3ed9004.png)

> I also noticed that the created Dockerfile has no exposed port. How is the debugger supposed to communicate with VSCode?

For Python debugging, we start `debugpy` in the container, with arguments to start the app itself. Then, `debugpy` makes a connection _out_ from the container to the debug adapter on the host. So a port open isn't necessary.

**Reply by bwateratmsft** [reply]
Still investigating further, but this might be a bug, with the same underlying issue as #3751.

**Reply by robs-nice99** [reply]
> Does the container stay running or immediately quit and/or get removed?

It stays running

> Does the "Python Debug Console" open?

no

**Reply by robs-nice99** [reply]
The "run and debug" "spinner keeps running also.
![image](https://user-images.githubusercontent.com/47054603/206701433-f4c01972-c767-496b-88bc-064b052b6c85.png)

**Reply by bwateratmsft** [reply]
Yeah, that's the exact same symptoms as #3751. We're working on a fix. In the meantime, it may work to roll back to Docker extension version 1.22.2. You can do so by running the command "Extensions: Install Specific Version of Extension..." in the command palette.

**Reply by robs-nice99** [reply]
I did the downgrade of the docker extension.
Besides the atlas error I also get this one.
![image](https://user-images.githubusercontent.com/47054603/207015348-1cb1a24f-7253-4fd3-8411-d6b9424bb4c7.png)
